module/os/linux/zfs/vdev_disk.c

```diff
@@ -671,35 +592,343 @@ vdev_bio_alloc(struct block_device *bdev, gfp_t gfp_mask,
+
+	vbio_return_abd(vbio);
+
+	kmem_free(vbio, sizeof (vbio_t) +
+	    sizeof (struct bio *) * vbio->vbio_max_bios);
+}
+
+static void
+vbio_put(vbio_t *vbio)
+{
+	if (atomic_dec_return(&vbio->vbio_ref) > 0)
+		return;
+
+	/*
+	 * This was the last reference, so the entire IO is completed. Clean
+	 * up and submit it for processing.
+	 */
+
+	/*
+	 * Get any data buf back to the original ABD, if necessary. We do
+	 * this now so we can submit the ZIO immediately and do remaining
+	 * cleanup after.
+	 */
+	vbio_return_abd(vbio);
+
+	zio_t *zio = vbio->vbio_zio;
+
+	/*
+	 * Set the overall error. If multiple BIOs returned an error, only the
+	 * first will be taken; the others are dropped (see
+	 * vdev_disk_io_rw_completion()). Its pretty much impossible for
+	 * multiple IOs to the same device to fail with different errors, so
+	 * there's no real risk.
+	 */
+	zio->io_error = vbio->vbio_error;
+	if (zio->io_error)
+		vdev_disk_error(zio);
+
+	/* All done, submit for processing */
+	zio_delay_interrupt(zio);
+
+	/* Finish cleanup */
+	vbio_free(vbio);
+}
+
+BIO_END_IO_PROTO(vdev_disk_io_rw_completion, bio, error)
+{
+	vbio_t *vbio = bio->bi_private;
+
+	if (vbio->vbio_error == 0) {
+#ifdef HAVE_1ARG_BIO_END_IO_T
+		vbio->vbio_error = BIO_END_IO_ERROR(bio);
+#else
+		if (error)
+			vbio->vbio_error = -(error);
+		else if (!test_bit(BIO_UPTODATE, &bio->bi_flags))
+			vbio->vbio_error = EIO;
+#endif
+	}
+
+	/* Drop this BIOs reference acquired by vbio_submit() */
+	vbio_put(vbio);
+}
+
+/*
+ * Iterator callback to count ABD pages and check their size & alignment.
+ *
+ * On Linux, each BIO segment can take a page pointer, and an offset+length of
+ * the data within that page. A page can be arbitrarily large ("compound"
+ * pages) but we still have to ensure the data portion is correctly sized and
+ * aligned to the natural page size, to ensure that if the kernel wants to
+ * split the BIO, the two halves will still be properly aligned.
+ */
+typedef struct {
+	uint_t	npages;
+	uint_t	end;
+} vdev_disk_check_pages_t;
+
+static int
+vdev_disk_check_pages_cb(void *buf, size_t len, void *priv)
+{
+	vdev_disk_check_pages_t *s = priv;
+
+	if (s->end != 0)
+		return (1);
+
+	while (len > 0) {
+		unsigned int off = offset_in_page(buf);
+
+		/* All pages after the first must start on offset 0 */
+		if (s->npages != 0 && off != 0)
+			return (1);
+
+		struct page *page = is_vmalloc_addr(buf) ?
+		    vmalloc_to_page(buf) : virt_to_page(buf);
+
+		size_t take = MIN(len, page_size(page)-off);
+
+		buf += take;
+		len -= take;
+
+		s->npages++;
+	}
+
+	/* Note if we're still in the middle of the last page */
+	s->end = offset_in_page(buf);
+	return (0);
+}
+
+/*
+ * Check if we can submit the pages in this ABD to the kernel as-is. Returns
+ * the number of pages, or 0 if it can't be submitted like this.
+ */
+static uint_t
+vdev_disk_check_pages(abd_t *abd)
+{
+	vdev_disk_check_pages_t s = {0};
+	if (abd_iterate_func(abd, 0, abd_get_size(abd),
```

Comment:
In FreeBSD similar code uses zio->io_size here. I am not completely sure it matches abd_get_size(abd) or is allowed to be smaller.